api/lib/domain/usecases/start-campaign-participation.js

```diff
@@ -2,13 +2,15 @@ const Assessment = require('../models/Assessment');
 
 const { NotFoundError } = require('../../domain/errors');
 
-module.exports = function startCampaignParticipation({ campaignParticipation, userId, campaignParticipationRepository, assessmentRepository, campaignRepository }) {
-  return _checkCampaignExists(campaignParticipation.campaignId, campaignRepository)
-    .then(() => _createSmartPlacementAssessment(userId, assessmentRepository))
-    .then((assessment) => _saveCampaignParticipation(campaignParticipation, assessment, userId, campaignParticipationRepository));
+module.exports = async function startCampaignParticipation({ campaignParticipation, userId, campaignParticipationRepository, assessmentRepository, campaignRepository }) {
+  await _checkCampaignExists(campaignParticipation.campaignId, campaignRepository);
+  const assessment = await _createSmartPlacementAssessment(userId, assessmentRepository);
+  const campaignParticipationCreated = await _saveCampaignParticipation(campaignParticipation, assessment, userId, campaignParticipationRepository);
+  await _updateAssessmentWithCampaignParticipation(assessmentRepository, assessment, campaignParticipationCreated);
```

Comment:
Je constate qu'on est obligé de jongler (création _assessment_, création _campaignParticipation_, mise à jour _assessment_) à cause de la référence circulaire introduite entre les tables `campaign-participations` et `assessments`.

J'ai l'impression qu'on pourrait se passer de cette référence circulaire… le modèle relationnel permet déjà de retrouver le(s) `assessments` attaché(s) à une `campaign-participation`. Ça me donne envie d'aller voir quelle quantité de code serait impactée par ce changement (ne plus avoir de `assessmendIt` dans `campaign-participations`).

Ça me fait aussi penser à nouveau que ce changement dans la relation entre _campaign participation_ et _assessment_ mérite une PR séparée.

Reply:
Retirer assessmentId de campaignParticipation me parait s'éloigner de ce qu'on voulait faire : retirer la responsabilité à l'assessment de gerer la campaignParticipation, mais que ce soit la campaignparticipation qui gère ses assessments ?

Reply:
Je crois qu'on est sur un exemple de divergence entre les modèles relationnel et objet (cf. http://blogs.tedneward.com/post/the-vietnam-of-computer-science/ ) : "la campaignparticipation qui gère ses assessments" peut être vrai/souhaitable côté objet, sans contradiction avec le fait que ce soit une colonne `assessments.campaignParticipationId` qui porte la relation avec `campaign-participations`.

Dans la base de données je préfère éviter de stocker une information de manière redondante, et/ou circulaire comme ici. Avec `assessments.campaignParticipationId` on mémorise bien la liste des _assessments_ associés à une _campaignParticipation_.

En partant de la _campaignParticipation_ on peut avoir besoin de choisir un des _assessments_ associés : par exemple pour la reprise, on voudra prendre le plus récent. Mais s'il y en a un plus ancien, il reste associé à cette participation et on pourrait utiliser cette information pour des statistiques ultérieures, par exemple.

En partant de l'_assessment_, on a besoin lors du `getNextChallenge` de retrouver le profil cible, et on le fait via la _campaignParticipation_. Mais actuellement pour retrouver cette _campaignParticipation_ on est obligé de faire une recherche par `assessmentId` dans la table `campaign-participations` (c'est ce que fait le `hasOne` dans `Assessment`), alors qu'avec `assessments.campaignParticipationId` on n'a qu'à suivre naturellement un pointeur.